### PR TITLE
Value representation rework

### PR DESCRIPTION
**Pull request by SeungheonOh**
superceeds #734, depends on #776

Rework of Data and Scott encoding and addition of SOP encoding. Scott encoding implementation is more or less same. However, PDataRecord/PDataSum is entirely deprecated

### CODE REVIEW

**Comment by SeungheonOh** [comment]
Golden is not updated intentionally. I'll update once I figure out all optimization

File: Plutarch/Repr/Internal.hs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+  All PEq struct =>
+  NP (Term s) struct ->
+  NP (Term s) struct ->
+  Term s PBool
+grecEq x y =
+  let
+    go :: [Term s PBool] -> Term s PBool
+    go [] = pconstant True
+    go [x'] = x'
+    go (x' : xs') = pif x' (go xs') (pconstant False)
```
Comment (SeungheonOh) [inline Plutarch/Repr/Internal.hs:81]: Small optimization here! because previously it was using `#&&` and `#&&` introduces unnecessary abs and app

File: Plutarch/Repr/SOP.hs [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    phoistAcyclic
+      ( plam $ \x' y' ->
+          pmatch x' $ \(PSOPRec (PRec x'')) ->
+            pmatch y' $ \(PSOPRec (PRec y'')) ->
+              grecEq x'' y''
+      )
+      # x
+      # y
+
+-- NOTE/TODO: Performance has not been tested for this. I feel like this would be terribly slow
```
Comment (SeungheonOh) [inline Plutarch/Repr/SOP.hs:65]: Though this still is slow, it's actually faster than original implementation. Because it doesn't use `#&&`

File: Plutarch/Builtin.hs [hunk 3]
```diff
@@ -1,706 +1,13 @@
...
-  PAsData,
-  pforgetData,
-  prememberData,
-  prememberData',
-  pserialiseData,
-  ppairDataBuiltin,
-  pchooseListBuiltin,
-  pchooseData,
-  PDataNewtype (..),
+  module Export,
```
Comment (kozross) [inline Plutarch/Builtin.hs:2]: Is this just a re-export module now? In that case, I think it should be removed. Re-export modules are confusing.

File: Plutarch/Builtin/Bool.hs [hunk 4]
```diff
@@ -18,21 +18,12 @@ module Plutarch.Builtin.Bool (
   pand',
   por',
   pcond,
+  ptrue,
+  pfalse,
```
Comment (kozross) [inline Plutarch/Builtin/Bool.hs:22]: Why do we need these?

File: Plutarch/Builtin/ByteString.hs [hunk 5]
```diff
@@ -1,112 +1,28 @@
...
-  punsafeIntegerToByte,
-
-  -- ** Other
-  phexByteStr,
-) where
-
-import Data.ByteString (ByteString)
+{-# OPTIONS_GHC -Wno-missing-export-lists #-}
+
+module Plutarch.Builtin.ByteString where
```
Comment (kozross) [inline Plutarch/Builtin/ByteString.hs:4]: This really should have explicit imports, and an explicit export list.

File: Plutarch/Builtin/ByteString.hs [hunk 6]
```diff
@@ -358,7 +192,7 @@ pallBS ::
   Term s ((PByte :--> PBool) :--> PByteString :--> PBool)
 pallBS = phoistAcyclic $ plam $ \p bs ->
   plet (plengthBS # bs) $ \len ->
-    go p len bs # 0
+    go p len bs # pconstantInteger 0
```
Comment (kozross) [inline Plutarch/Builtin/ByteString.hs:195]: Why this change? Cycle break?

File: Plutarch/Builtin/Data.hs [hunk 7]
```diff
@@ -0,0 +1,109 @@
+{-
+BuiltinPair and BuiltinList should go into their own module !!
+-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Plutarch.Builtin.Data (
```
Comment (kozross) [inline Plutarch/Builtin/Data.hs:7]: Since these have all been moved, `@since`s should be added too.

File: Plutarch/Convert.hs [hunk 8]
```diff
@@ -15,21 +15,7 @@ module Plutarch.Convert (
...
-import Plutarch.Internal.Other (pto)
-import Plutarch.Internal.PLam (plam)
-import Plutarch.Internal.PlutusType (
-  DPTStrat,
-  DerivePlutusType,
-  PlutusType,
-  pcon,
- )
-import Plutarch.Internal.Term (S, Term, (#), (:-->))
+import Plutarch.Internal (DPTStrat, DerivePlutusType, PBool (PFalse, PTrue), PByteString, PEq, PInteger, POrd, PlutusType, PlutusTypeNewtype, S, Term, pcon, plam, pto, (#), (:-->))
```
Comment (kozross) [inline Plutarch/Convert.hs:18]: Could we roll this entire module in with `PByteString`? I think the deps should be breakable.

File: Plutarch/FFI.hs [hunk 9]
```diff
@@ -31,47 +31,11 @@ import Generics.SOP.Type.Metadata (
...
-  compile,
-  pdelay,
-  pforce,
-  phoistAcyclic,
-  (#),
-  (:-->),
- )
-import Plutarch.Internal.Witness (witness)
-import Plutarch.List (PList, PListLike (PElemConstraint, pcons, pelimList, pnil), pconvertLists, plistEquals)
+
```
Comment (kozross) [inline Plutarch/FFI.hs:34]: Does anything actually _use_ the FFI module?

File: Plutarch/Internal.hs [hunk 10]
```diff
@@ -0,0 +1,30 @@
+module Plutarch.Internal (
```
Comment (kozross) [inline Plutarch/Internal.hs:1]: Again, re-export modules are bad, kill it.

File: Plutarch/Internal/Eq.hs [hunk 11]
```diff
@@ -1,7 +1,14 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE UndecidableInstances #-}
+
 module Plutarch.Internal.Eq (
   PEq (..),
+  gpeq,
```
Comment (kozross) [inline Plutarch/Internal/Eq.hs:6]: Why do we need to expose this?

**Review by kozross — CHANGES_REQUESTED** [review]
In general:

* Moved identifiers need `@since`s
* Re-export modules (other than `Prelude`) shouldn't exist
* Explicit exports and explicit imports everywhere

Also, has this affected the publically-facing interface in any way? If so, we should changelog it.

File: Plutarch/Internal/Eq.hs (hunk 11 above)
Comment (SeungheonOh) [reply Plutarch/Internal/Eq.hs:6]: This will be killed anyways on #767

File: Plutarch/FFI.hs (hunk 9 above)
Comment (SeungheonOh) [reply Plutarch/FFI.hs:34]: Not really. might be nice to deprecate this, but not on this PR

File: Plutarch/Convert.hs (hunk 8 above)
Comment (SeungheonOh) [reply Plutarch/Convert.hs:18]: yes possibly. this PR's gone extremely big and I was very fed up near end. More can be done definitely. This is good enough for #767

File: Plutarch/Builtin/Bool.hs (hunk 4 above)
Comment (SeungheonOh) [reply Plutarch/Builtin/Bool.hs:22]: we can't use `pconstant` in Builtin submodules. Having extra module solely for utility functions seemed too annoying

**Comment by SeungheonOh** [comment]
@kozross Once you made changes(or reverts), I'll push PLiftable updates.

File: Plutarch/Convert.hs (hunk 8 above)
Comment (kozross) [reply Plutarch/Convert.hs:18]: I'll break it into another ticket.

File: Plutarch/FFI.hs (hunk 9 above)
Comment (kozross) [reply Plutarch/FFI.hs:34]: Will make a spinoff ticket then.

**Review by kozross — APPROVED** [review]
